keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2164 @@
+Pod may run. The downside is that pod scheduling is less flexible.
+
+For delayed allocation, a node is selected tentatively by the scheduler
+and driver(s) try to allocate their resources for that node. If that
+succeeds, the Pod can get scheduled. If it fails, the scheduler must
+determine whether some other node fits the requirements and if so,
+request allocation again. If no node fits because some resources were
+already allocated for a node and are only usable there, then those
+resources must be released and then get allocated elsewhere.
+
+This is a summary of the necessary [kube-scheduler changes](#kube-scheduler) in
+pseudo-code:
+
+```
+while <pod needs to be scheduled> {
+  <choose a node, considering potential availability for those resources
+   which are not allocated yet and the hard constraints for those which are>
+  if <no node fits the pod> {
+    if <at least one resource
+            is allocated and unused,
+            uses delayed allocation, and
+            was not available on a node> {
+      <randomly pick one of those resources and tell resource driver to free it>
+    }
+  } else if <all resources allocated> {
+    <schedule pod onto node>
+  } else if <some unallocated resource uses delayed allocation> {
+    <tell resource driver to allocate for the chosen node>
+  }
+}
+```
+
+### Resource allocation and usage flow
+
+The following steps shows how resource allocation works for a resource that
+gets defined inline in a Pod. Several of these steps may fail without changing
+the system state. They then must be retried until they succeed or something
+else changes in the system, like for example deleting objects.
+
+* **user** creates Pod with inline ResourceClaimTemplate
+* **resource claim controller** checks ResourceClaimTemplate and ResourceClass,
+  then creates ResourceClaim with Pod as owner
+* if *immediate allocation*:
+  * **resource driver** adds finalizer to claim to prevent deletion -> allocation in progress
+  * **resource driver** finishes allocation, sets `Allocation` -> claim ready for use by any pod
+* if *pod is pending*:
+  * **scheduler** filters nodes
+  * if *delayed allocation and resource not allocated yet*:
+    * if *at least one node fits pod*:
+      * **scheduler** picks one node, sets `SuitableNodes=<all nodes that fit the pod>`, `SelectedNode=<the chosen node>` and `SelectedUser=<the pod being scheduled>`
+      * if *resource is available for `SelectedNode`*:
+        * **resource driver** adds finalizer to claim to prevent deletion -> allocation in progress
+        * **resource driver** finishes allocation, sets `Allocation` and the
+          intended user in `ReservedFor` -> claim ready for use and reserved
+          for the pod
+      * else *scheduler needs to know that it must avoid this and possibly other nodes*:
+        * **resource driver** sets `UnsuitableNodes` -> next attempt by scheduler has more information and is more likely to succeed
```

Comment:
Having a node selector in the class is a good first-order filter, but may not be sufficient in all cases?

After digesting this more, I feel like, if we're going to do this sort of negotiation, it shouldn't be bolted onto the ResourceClaim object

Reply:
> Would the resource driver write all those pods names that don't have the label here?

A smart driver can avoid listing all the not-maching ones by setting DriverSchedulingStatus.SuitableNodes instead, which is a NodeSelector. The API as specified allows that, but I hadn't thought about this particular case, so the KEP doesn't discuss it.

But this problem already occurs when the scheduler puts nodes that don't have the label into PotentialNodes. That list is much larger than it should be.

> I think ResourceClass should at least have a NodeSelector that scheduler and clusterautoscaler can use beforehand. 

I like that idea.

> And this should be copied to the DriverSchedulingStatus at creation time 

That shouldn't be necessary: the ResourceClaim references the ResourceClass and the value from there can be used as additional filter criteria for unallocated ResourceClaims. A ResourceClass is required for allocation, not having one is an error.

Not copying avoids the problem of having stale information when a ResourceClass gets replaced or updated.

> maybe there should be a selector in the ResourceClaimSpec as well.

Putting it into the spec implies that the user chooses. Perhaps that is useful for some special cases, but we would need to investigate use cases for it first.

So to recap, DriverSchedulingStatus.SuitableNodes serves the same purpose as what you suggested (restrict potential nodes via labels), it just gets set differently (after ResourceClaim creation by the driver instead of before creation by the admin).

I like your proposal better, let me add it. It is slightly more work for the admin, but that's worth it.

DriverSchedulingStatus.SuitableNodes then can be removed, in particular...

> After digesting this more, I feel like, if we're going to do this sort of negotiation, it shouldn't be bolted onto the ResourceClaim object

... when we move PotentialNodes/UnsuitableNodes out of ResourceClaimStatus. Then the new object (`ResourceClaimScheduling`?) is only needed when the scheduler initiates the negotiation.

I'm okay with doing this via a separate object. Conceptually it feels cleaner. Performance might be worse (cannot create new object and set SelectedNode in a single API call) compared to the current proposal, but if performance is a concern, then the cluster probably should be configured to do node filtering via the scheduler extender mechanism.

Reply:
I want to be respectful of the work already put into this, but I also want to be really careful on something this "sophisticated".

If you zoom out we have, abstractly:

* The scheduler can see from a pending pod that not all of its dependencies are satisfied
* To schedule it must satisfy them
* To satisfy them, scheduler has to pull information and opinions from N independent actors
* These actors need a place and a protocol by which to compare notes and share information
* We know this is true for late-allocation volumes and resource claims today
* We know that our crystal ball is terrible at predicting the future

Ergo, attaching this negotiation to ResourceClaim seems highly suspect.  In absolute terms, performance might be worse than doing it in a single API resource, but we can't let the spectre of perf prematurely force our hands into terrible design choices.

That said, scheduler is NOT my primary domain, so this really really has to be something they agree with.

Reply:
I had already switched to a separate object (`ResourceClaimScheduling`) but let me go further and introduce a single `PodScheduling` object. You are making a good point about "these actors need **a** place" (emphasis mine) and from an API perspective a single object related to the pod is cleaner and truly independent of any particular `ResourceClaim`.